### ISSUE TITLE
issue: isDirty & dirtyFields are incorrect when a form contains a Controller-wrapped input with React.StrictMode and React 18

### ISSUE THREAD

**Issue by cschwebk** [react-hook-form/react-hook-form#10045]
### Version Number

7.43.4

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-defaultvalues-forked-ym0gmz?file=/src/index.js

### Steps to reproduce

1. View the Codesandbox link and open the console.
1. Notice two inputs, one directly registered (`firstName`), one wrapped in a `Controller` (`lastName`). 
1. Notice 4 log entries for `defaultValues`, 2 with `firstName` and `lastName`, followed by 2 with only `firstName`.
1. Modify the "First Name" input.
1. See that `isDirty` is now `true`, that `firstName` is now listed in `dirtyFields`
1. Revert the changes to the "First Name" input
1. See that `isDirty` is still `true`, and `firstName` is not listed in `dirtyFields`.
1. Modify the "Last Name" input.
1. See that `isDirty` is still `true`, and `lastName` is now listed in `dirtyFields`.
1. Revert the changes to the "Last Name" input.
1. Notice that no new log entries are produced when changing the "Last Name" input back to the default value.
1. Modify the "First Name" input.
1. Notice that `lastName` is still listed in `dirtyFields`.
1. Comment out the call to `root.render` that contains `React.StrictMode` and uncomment the call to `root.render` without `React.StrictMode`.
1. Verify that the form correctly tracks `isDirty` and `dirtyField` form state.

### Expected behaviour

For an application using React 18 and `React.StrictMode`, a form containing an input wrapped in a Controller should continue to properly update the `isDirty` and `dirtyFields` form state when the form input values change.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
This is a design limitation with `shouldUnregister`, as we set input to unmounted but `useEffect` invoked twice making this almost impossible and diffcult track the input unmount state.

**Comment by cschwebk** [comment]
Thanks for your reply @bluebill1049. I had made so many changes to that codesandbox that I forgot I still had `shouldUnregister` on only the `Controller`. I updated the sandbox to include `shouldUnregister` in the `register` call `options`, and the functionality of the `firstName` input does not have the same issues as the `lastName` input. Is the design limitation strictly related to the combination of React 18, `React.StrictMode`, `Controller` and `shouldUnregister`? 

Perhaps the note on `shouldUnregister` in the documentation could be expanded to also warn about this interaction.
